### PR TITLE
Import protocol modal with preview [SCI-3564] & [SCI-3568]

### PR DESCRIPTION
**Pull request by urbanrotnik**
Jira ticket: [SCI-3568](https://biosistemika.atlassian.net/browse/SCI-3568) [SCI-3564](https://biosistemika.atlassian.net/browse/SCI-3564)

### What was done
Preview modal, JS for handling import and build Actions, serializing JSON (no sucess with ActiveModelSerializer), build preview

#### ToDo:
Will be defined on todays groomig

### CODE REVIEW

File: app/controllers/external_protocols_controller.rb [hunk 1]
```diff
@@ -48,13 +48,21 @@ def new
...
+    @protocol = p
     if service_call.succeed?
       render json: {
         html: render_to_string(
           partial: 'protocol_importers/import_form.html.erb',
-          locals: { protocol: service_call.built_protocol }
-        )
+          locals: { protocol: p, steps_json: service_call.serialized_steps }
+        ),
+        title: "Import protocol - #{p.name}",
```
Comment (jbargu) [inline app/controllers/external_protocols_controller.rb]: I18

File: app/controllers/external_protocols_controller.rb [hunk 2]
```diff
@@ -48,13 +48,21 @@ def new
       user_id: current_user.id,
       team_id: @team.id
     )
+    p = service_call.built_protocol
+    p.valid? # Get validations errors here
 
+    @protocol = p
```
Comment (jbargu) [inline app/controllers/external_protocols_controller.rb]: We could just assing this above at `service_call.built_protocol.

File: app/assets/stylesheets/protocols/preview_modal.scss [hunk 3]
```diff
@@ -14,4 +14,8 @@
       min-height: initial;
     }
   }
+
+  .general-eror {
```
Comment (jbargu) [inline app/assets/stylesheets/protocols/preview_modal.scss]: Typo `erRor`.

File: app/controllers/external_protocols_controller.rb [hunk 4]
```diff
@@ -64,14 +72,17 @@ def new
   # POST import_external_protocol
   def create
     service_call = ProtocolImporters::ImportProtocolService.call(
-      protocol_params: create_params[:protocol_params],
-      steps_params: create_params[:steps_paramas],
+      protocol_params: create_protocol_params,
+      steps_params: JSON.parse(create_steps_params[:steps]),
```
Comment (jbargu) [inline app/controllers/external_protocols_controller.rb]: We probably want to parse this inside the service to catch errors if parse fails here.

File: app/assets/javascripts/sitewide/external_protocols.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      });
+      e.preventDefault();
+      return false;
+    });
+  }
+
+  function initFormSubmits() {
+    var modal = $('#protocol-preview-modal');
+    modal.on('click', 'button[data-action=import_protocol]', function() {
+      // console.log('click');
```
Comment (jbargu) [inline app/assets/javascripts/sitewide/external_protocols.js]: Remove

**Comment by jbargu** [comment]
From my side it looks more than ok, really amazing job! Although you have to fix the tests.

**Comment by urbanrotnik** [comment]
> image
> You're probably aware of this.
> 
> And additionally some problems with bigger text probably:
> image

Yeah... I know for this. "Advanced sanitization"  should be solved in a separate task. Also FE bug maybe. I just reused the existing design.

File: app/assets/javascripts/sitewide/external_protocols.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  function showFormErrors(modal, errors) {
+    resetFormErrors(modal);
+
+    // AR valdiation errors
+    Object.keys(errors.protocol).forEach(function(key) {
+      var input = modal.find('#protocol_' + key);
+      var msg;
+      msg = key.charAt(0).toUpperCase() + key.slice(1) + ': ' + errors.protocol[key].join(', ');
```
[the remark this answers is not in the document]
Comment (jbargu) [reply app/assets/javascripts/sitewide/external_protocols.js]: This is a question of a whole app how to treat validation errors and as such it's probably not applicable at this moment.

[the remark this answers is not in the document]
Comment (urbanrotnik) [reply app/assets/javascripts/sitewide/external_protocols.js]: Hmm... Yes, I agree, but maybe not needed for this task. We should translate/define all standard errors/validations than I think?

File: app/views/steps/attachments/_preview_list.html.erb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+</div>
+<div class="col-xs-12 attachments-actions">
+
+  <div class="title">
+    <h4>
+      <%= t('protocols.steps.files', count: attachments.size) %>
+    </h4>
+  </div>
+  <div>
+    <div class="attachemnts-header pull-right">
```
Comment (jbargu) [inline app/views/steps/attachments/_preview_list.html.erb]: Typo

**Pull request merged**